pkg/pma/pma.c

```diff
@@ -0,0 +1,816 @@
+                    wal_dir,
+                    strerror(err));
+            goto fail;
+        }
+        free(heap_file_copy);
+    }
+
+    if (pma->heap_fd != -1) {
+        assert(wal);
+        if (_append_dirty_pages(wal, pma->heap_start, false, pma, heap_len)
+            == -1)
+        {
+            err = ECANCELED;
+            fprintf(stderr,
+                    "pma: failed to sync heap changes to %s\r\n",
+                    pma->heap_file);
+            goto fail;
+        }
+    }
+
+    if (pma->stack_fd != -1) {
+        assert(wal);
+        if (_append_dirty_pages(wal, pma->stack_start, true, pma, stack_len)
+            == -1)
+        {
+            err = errno;
+            fprintf(stderr,
+                    "pma: failed to sync stack changes to %s: %s\r\n",
+                    pma->stack_file,
+                    strerror(err));
+            goto fail;
+        }
+    }
+
+    // Sync and apply the WAL if the backing files are non-NULL.
+    if (wal) {
+        if (wal_sync(wal) == -1) {
+            err = errno;
+            fprintf(stderr,
+                    "pma: failed to sync wal for %s: %s\r\n",
+                    wal_dir,
+                    strerror(err));
+            goto fail;
+        }
+
+        if (wal_apply(wal, pma->heap_fd, pma->stack_fd) == -1) {
+            err = errno;
+            fprintf(stderr,
+                    "pma: failed to apply wal to %s: %s\r\n",
+                    wal_dir,
+                    strerror(err));
+            goto fail;
+        }
+
+        wal_destroy(wal);
+    }
+
+    // Remap heap.
+    if (_map_file(pma->heap_fd, pma->heap_start, false, pma, &pma->heap_len)
+        == -1)
+    {
+        err = errno;
+        fprintf(stderr,
+                "pma: failed to remap %s: %s\r\n",
+                pma->heap_file,
+                strerror(err));
+        goto fail;
+    }
+
+    // Remap stack.
+    if (_map_file(pma->stack_fd, pma->stack_start, true, pma, &pma->stack_len)
+        == -1)
+    {
+        err = errno;
+        fprintf(stderr,
+                "pma: failed to remap %s: %s\r\n",
+                pma->stack_file,
+                strerror(err));
+        munmap(pma->heap_start, pma->heap_len);
+        close(pma->heap_fd);
+        pma->heap_fd = -1;
+        goto fail;
+    }
+
+    // Return ephemeral memory.
+    //
+    // To verify correct behavior, watch memory usage after the command:
+    //
+    //   =/  a  (bex (bex 30))  ~
+    //
+    // This is likely fast enough to do after every event, at least if
+    // we use MADV_FREE, which is lazy.  We use the strict MADV_DONTNEED
+    // to make it easier to observe its behavior.
+    char  *unmap_start = (char *)pma->heap_start + pma->heap_len;
+    size_t unmap_len   = _total_len(pma) - (pma->heap_len + pma->stack_len);
+    assert(unmap_len % kPageSz == 0);
+    if (madvise(unmap_start, unmap_len, MADV_DONTNEED) == -1) {
```

Comment:
is `unmap_start` always guaranteed page-aligned? Additionally, is `unmap_len` guaranteed page-aligned? @philipcmonk

Reply:
I suppose `unmap_start` is because it should be `U3_OS_LoomBase`. This makes the the `heap_start` naming somewhat confusing because that isn't actually the start of our heap. `U3_OS_LoomBase + (2 * sizeof(c3_w))` is.

I don't think `unmap_len` is though. We should round that down otherwise, is corruption (dropping anon memory that we actually need) possible?

Reply:
Oh duh, ignore the above. I see the assert now